### PR TITLE
Configure challenges api

### PR DESCRIPTION
**Pull request by alinagoal**
- Created necessary api for challenges:
 - DTO,controller, service, module
 - data seeding
 -  swagger docs
 -  code docs

### CODE REVIEW

File: apps/server/src/challenges/challenges.controller.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import {
+  Body,
```
Comment (SamipSthaLF) [inline apps/server/src/challenges/challenges.controller.ts:2]: ladder.

File: apps/server/src/challenges/challenges.controller.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import { Roles } from 'src/decorator/roles.decorator';
+
+import { RoleEnum } from 'src/roles/seed/role.enum';
+
+import { ChallengesDto } from './dto/challenges.dto';
+
+import { ChallengesService } from './challenges.service';
+
+@Controller('challenges')
```
Comment (SamipSthaLF) [inline apps/server/src/challenges/challenges.controller.ts:22]: swagger

File: apps/server/src/challenges/seed/challenges.seed.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { Injectable } from '@nestjs/common';
+
+import { ChallengesService } from '../challenges.service';
+@Injectable()
+export class ChallengesSeed {
+  constructor(private readonly challengesService: ChallengesService) {}
+
+  async seed() {
```
Comment (SamipSthaLF) [inline apps/server/src/challenges/seed/challenges.seed.ts:8]: code doc.

File: apps/server/src/challenges/seed/challenges.seed.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { Injectable } from '@nestjs/common';
+
+import { ChallengesService } from '../challenges.service';
+@Injectable()
```
Comment (SamipSthaLF) [inline apps/server/src/challenges/seed/challenges.seed.ts:4]: leave space before.

File: apps/server/src/main.ts [hunk 5]
```diff
@@ -20,6 +21,8 @@ async function bootstrap() {
   await roleSeed.seed();
   const userSeed = app.get(UserSeed);
   await userSeed.seed();
+  const challengesSeed = app.get(ChallengesSeed);
```
Comment (SamipSthaLF) [inline apps/server/src/main.ts:24]: space out and group declarations.

File: apps/server/src/user/entities/user.entity.ts [hunk 6]
```diff
@@ -1,16 +1,11 @@
 import { Role } from 'src/roles/entities/role.entity';
 
 import {
-  AfterInsert,
-  AfterRemove,
-  AfterUpdate,
   BaseEntity,
```
Comment (SamipSthaLF) [inline apps/server/src/user/entities/user.entity.ts:4]: ladder.

File: apps/server/src/userroles/userroles.service.ts [hunk 7]
```diff
@@ -31,7 +31,7 @@ export class UserrolesService {
       );
     }
 
-    var roleList: Role[] = [];
+    const roleList: Role[] = [];
```
Comment (SamipSthaLF) [inline apps/server/src/userroles/userroles.service.ts:34]: leave space after declarations.

File: src/testnew/testnew.controller.spec.ts [hunk 8]
```diff
@@ -1,8 +1,8 @@
-import { Test, TestingModule } from '@nestjs/testing';
-import { TestnewController } from './testnew.controller';
-import { TestnewService } from './testnew.service';
+import { Test, TestingModule } from "@nestjs/testing";
+import { TestnewController } from "./testnew.controller";
+import { TestnewService } from "./testnew.service";
```
Comment (SamipSthaLF) [inline src/testnew/testnew.controller.spec.ts:3]: group and ladder imports.

File: src/testnew/testnew.controller.ts [hunk 9]
```diff
@@ -1,9 +1,17 @@
-import { Controller, Get, Post, Body, Patch, Param, Delete } from '@nestjs/common';
-import { TestnewService } from './testnew.service';
-import { CreateTestnewDto } from './dto/create-testnew.dto';
-import { UpdateTestnewDto } from './dto/update-testnew.dto';
+import {
+  Controller,
```
Comment (SamipSthaLF) [inline src/testnew/testnew.controller.ts:2]: ladder

File: src/testnew/testnew.controller.ts [hunk 10]
```diff
@@ -1,9 +1,17 @@
...
+  Patch,
+  Param,
+  Delete,
+} from "@nestjs/common";
+import { TestnewService } from "./testnew.service";
+import { CreateTestnewDto } from "./dto/create-testnew.dto";
+import { UpdateTestnewDto } from "./dto/update-testnew.dto";
 
-@Controller('testnew')
+@Controller("testnew")
```
Comment (SamipSthaLF) [inline src/testnew/testnew.controller.ts:14]: swagger tag.

File: src/testnew/testnew.service.ts [hunk 11]
```diff
@@ -1,11 +1,11 @@
-import { Injectable } from '@nestjs/common';
-import { CreateTestnewDto } from './dto/create-testnew.dto';
-import { UpdateTestnewDto } from './dto/update-testnew.dto';
+import { Injectable } from "@nestjs/common";
+import { CreateTestnewDto } from "./dto/create-testnew.dto";
+import { UpdateTestnewDto } from "./dto/update-testnew.dto";
 
 @Injectable()
 export class TestnewService {
   create(createTestnewDto: CreateTestnewDto) {
```
Comment (SamipSthaLF) [inline src/testnew/testnew.service.ts:7]: add code doc.

File: apps/server/src/challenges/seed/challenges.seed.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+export class ChallengesSeed {
+  constructor(private readonly challengesService: ChallengesService) {}
+
+  async seed() {
+    const existingChallenges = await this.challengesService.findAll();
+    if (existingChallenges.length === 0) {
```
Comment (SamipSthaLF) [inline apps/server/src/challenges/seed/challenges.seed.ts:10]: add space after declaratons.

File: apps/server/src/challenges/dto/challenges.dto.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
+
+export type visibilityOptions = 'private' | 'public';
+export class ChallengesDto {
```
Comment (SamipSthaLF) [inline apps/server/src/challenges/dto/challenges.dto.ts:4]: add space between exports.

File: apps/server/src/challenges/challenges.controller.ts (hunk 1 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:2]: Done

File: apps/server/src/challenges/seed/challenges.seed.ts (hunk 4 above)
Comment (alinagoal) [reply apps/server/src/challenges/seed/challenges.seed.ts:4]: Done

File: apps/server/src/main.ts (hunk 5 above)
Comment (alinagoal) [reply apps/server/src/main.ts:24]: Done

File: apps/server/src/challenges/entities/challenges.entity.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
+import {
```
[the remark this answers is not in the document]
Comment (alinagoal) [reply apps/server/src/challenges/entities/challenges.entity.ts:1]: Done

File: apps/server/src/challenges/challenges.controller.ts (hunk 2 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:22]: Done

File: apps/server/src/challenges/seed/challenges.seed.ts (hunk 3 above)
Comment (alinagoal) [reply apps/server/src/challenges/seed/challenges.seed.ts:13]: Done

File: apps/server/src/challenges/challenges.controller.ts (hunk 2 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:29]: Done

File: apps/server/src/challenges/seed/challenges.seed.ts (hunk 4 above)
Comment (alinagoal) [reply apps/server/src/challenges/seed/challenges.seed.ts:5]: Done

File: apps/server/src/challenges/dto/challenges.dto.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+  @ApiProperty() public title: string;
+  @ApiPropertyOptional() public description?: string;
+  @ApiProperty() public points: number;
```
[the remark this answers is not in the document]
Comment (alinagoal) [reply apps/server/src/challenges/dto/challenges.dto.ts]: Done

File: src/testnew/testnew.service.ts [hunk 16]
```diff
@@ -1,11 +1,11 @@
-import { Injectable } from '@nestjs/common';
-import { CreateTestnewDto } from './dto/create-testnew.dto';
-import { UpdateTestnewDto } from './dto/update-testnew.dto';
+import { Injectable } from "@nestjs/common";
+import { CreateTestnewDto } from "./dto/create-testnew.dto";
```
[the remark this answers is not in the document]
Comment (alinagoal) [reply src/testnew/testnew.service.ts:2]: Done

File: apps/server/src/app.module.ts [hunk 17]
```diff
@@ -6,20 +6,21 @@ import { ConfigModule, ConfigService } from '@nestjs/config';
...
-import { AuthModule } from '@/auth/auth.module';
+import { AuthModule } from './auth/auth.module';
 
-import { UserModule } from '@/user/user.module';
+import { UserModule } from './user/user.module';
+
+import { RolesModule } from '@/roles/roles.module';
+
+import { UserrolesModule } from './userroles/userroles.module';
+import { ChallengesModule } from './challenges/challenges.module';
```
Comment (AsminShresthaLF) [inline apps/server/src/app.module.ts:16]: Let's use @ alias.

File: apps/server/src/challenges/challenges.service.ts [hunk 18]
```diff
@@ -0,0 +1,65 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+
+import { Repository } from 'typeorm';
+
+import { Challenge } from './entities/challenges.entity';
+import { CreateChallengeDto } from './dto/challenges.dto';
```
Comment (AsminShresthaLF) [inline apps/server/src/challenges/challenges.service.ts:7]: Alias

File: apps/server/src/app.module.ts (hunk 17 above)
Comment (alinagoal) [reply apps/server/src/app.module.ts:16]: Done

File: apps/server/src/challenges/challenges.service.ts (hunk 18 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.service.ts:7]: Done

File: apps/server/src/app.module.ts [hunk 19]
```diff
@@ -14,6 +14,7 @@ import { AppController } from '@/app.controller';
 import getDBConfig from '@/config/typeorm.config';
 
 import { JwtStrategy } from '@filter/jwt.strategy';
+import { ChallengesModule } from './challenges/challenges.module';
```
Comment (yachujoshi-lf) [inline apps/server/src/app.module.ts:17]: Let's use import aliases.

File: apps/server/src/challenges/challenges.controller.ts [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+  Post,
+  Param,
+  Patch,
+  Delete,
+  HttpCode,
+  Controller,
+  HttpStatus,
+} from '@nestjs/common';
+
+import { Roles } from 'src/decorator/roles.decorator';
```
Comment (yachujoshi-lf) [inline apps/server/src/challenges/challenges.controller.ts:19]: .

File: apps/server/src/challenges/challenges.controller.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+  constructor(private readonly challengesService: ChallengesService) {}
+
+  @Post()
+  @Roles(RoleEnum.ADMIN)
+  @ApiCreatedResponse({
+    description: 'Created a new challenge',
+    type: Challenge,
+  })
+  @ApiBadRequestResponse({ description: 'Challenge cannot be created' })
+  @HttpCode(HttpStatus.CREATED)
```
Comment (yachujoshi-lf) [inline apps/server/src/challenges/challenges.controller.ts:45]: No need to specify `HttpCode`. Nest will resolve this for us.

File: apps/server/src/challenges/challenges.controller.ts [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  @Get()
+  @Roles(RoleEnum.ADMIN, RoleEnum.REVIEWER, RoleEnum.USER)
+  @ApiCreatedResponse({
+    description: 'Fetched all challenges',
+    type: Challenge,
+  })
+  @ApiBadRequestResponse({ description: 'Challenge cannot be fetched' })
+  @HttpCode(HttpStatus.OK)
```
Comment (yachujoshi-lf) [inline apps/server/src/challenges/challenges.controller.ts:57]: .

File: apps/server/src/challenges/challenges.controller.ts [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+
+  @Get(':id')
+  @Roles(RoleEnum.ADMIN, RoleEnum.REVIEWER, RoleEnum.USER)
+  @ApiCreatedResponse({
+    description: 'Fetched challenge',
+    type: Challenge,
+  })
+  @ApiBadRequestResponse({ description: 'Challenge cannot be fetched' })
+  @HttpCode(HttpStatus.OK)
+  public async getById(@Param('id') id: number) {
```
Comment (yachujoshi-lf) [inline apps/server/src/challenges/challenges.controller.ts:70]: No need to mark controller methods as `public`

File: apps/server/src/challenges/challenges.service.ts [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+    @InjectRepository(Challenge)
+    private challengeRepository: Repository<Challenge>,
+  ) {}
+
+  /**
+   * Create a new challenge.
+   * @param {CreateChallengeDto} payload - The payload containing challenge information.
+   * @returns {Promise<ChallengeEntity>} - The created challenge entity.
+   */
+  public async create(payload: CreateChallengeDto) {
```
Comment (yachujoshi-lf) [inline apps/server/src/challenges/challenges.service.ts:21]: We don't need to mark these as `public` as well.

File: apps/server/src/main.ts [hunk 25]
```diff
@@ -2,24 +2,34 @@ import { NestFactory, Reflector } from '@nestjs/core';
...
    * Seeds the user to <SEED_USER_EMAIL]> if not present and assign default role to the user
+   * Seeds the challenges table
    */
   const userSeed = app.get(UserSeed);
 
   //seed user
   await userSeed.seed();
 
+  const challengesSeed = app.get(ChallengesSeed);
+  await challengesSeed.seed();
```
Comment (yachujoshi-lf) [inline apps/server/src/main.ts:31]: Let's refactor this logic to a separate function. And let's use `Promise.all()` for this.

File: apps/server/swagger.config.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
+import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';
+
+export function setupSwagger(app: any) {
```
Comment (yachujoshi-lf) [inline apps/server/swagger.config.ts:3]: Let's avoid `any` type.

File: apps/server/src/challenges/dto/challenges.dto.ts [hunk 27]
```diff
@@ -0,0 +1,22 @@
+import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
+
+export type visibilityOptions = 'private' | 'public';
+
+export class CreateChallengeDto {
```
Comment (yachujoshi-lf) [inline apps/server/src/challenges/dto/challenges.dto.ts:5]: This should be in `create-challenges.dto.ts` file.

File: apps/server/src/challenges/dto/challenges.dto.ts [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+  @ApiProperty({ default: 'public' }) public privacy: visibilityOptions;
+
+  @ApiProperty({ default: 1 }) public author_id: number;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export interface UpdateChallengeDto extends Partial<CreateChallengeDto> {}
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export class UpdateChallengeDto {}
```
Comment (yachujoshi-lf) [inline apps/server/src/challenges/dto/challenges.dto.ts:22]: And this should be in `update-challenge.dto.ts` file.

File: apps/server/src/app.module.ts (hunk 19 above)
Comment (alinagoal) [reply apps/server/src/app.module.ts:17]: Done

File: apps/server/src/challenges/challenges.controller.ts (hunk 20 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:19]: Done

File: apps/server/src/challenges/challenges.controller.ts (hunk 21 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:45]: Hmm, I used this for swagger api. Is this not needed there as well?

Comment (yachujoshi-lf) [reply apps/server/src/challenges/challenges.controller.ts:45]: `@ApiOkResponse()` for 200 and `@ApiCreatedResponse` for 201

File: apps/server/src/challenges/challenges.service.ts (hunk 24 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.service.ts:21]: Done

File: apps/server/src/challenges/challenges.controller.ts (hunk 23 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:70]: Done

File: apps/server/swagger.config.ts (hunk 26 above)
Comment (alinagoal) [reply apps/server/swagger.config.ts:3]: Done

File: apps/server/src/challenges/dto/challenges.dto.ts (hunk 27 above)
Comment (alinagoal) [reply apps/server/src/challenges/dto/challenges.dto.ts:5]: Done

File: apps/server/src/challenges/dto/challenges.dto.ts (hunk 28 above)
Comment (alinagoal) [reply apps/server/src/challenges/dto/challenges.dto.ts:22]: Done

File: apps/server/src/challenges/challenges.controller.ts (hunk 21 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:45]: okay

File: apps/server/src/main.ts (hunk 25 above)
Comment (alinagoal) [reply apps/server/src/main.ts:31]: Done

File: apps/server/src/challenges/challenges.controller.ts (hunk 22 above)
Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.ts:57]: Done

File: apps/server/src/challenges/challenges.controller.spec.ts [hunk 29]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { ChallengesController } from '@/challenges/challenges.controller';
```
Comment (AsminShresthaLF) [inline apps/server/src/challenges/challenges.controller.spec.ts:2]: Looks like we need spacing here.

Comment (alinagoal) [reply apps/server/src/challenges/challenges.controller.spec.ts]: Done

**Review by SamipSthaLF — APPROVED** [review]

File: apps/server/src/main.ts [hunk 30]
```diff
@@ -2,24 +2,36 @@ import { NestFactory, Reflector } from '@nestjs/core';
...
+   * Seeds the challenges table
    */
   const userSeed = app.get(UserSeed);
 
-  //seed user
-  await userSeed.seed();
+  const challengesSeed = app.get(ChallengesSeed);
 
-  await app.listen(3001);
+  Promise.all([userSeed.seed(), challengesSeed.seed()]);
```
Comment (yachujoshi-lf) [inline apps/server/src/main.ts:36]: `await Promise.all([userSeed.seed(), challengesSeed.seed()]);`

**Review by yachujoshi-lf — APPROVED** [review]

**Pull request merged**